pandas/core/indexes/base.py

```diff
@@ -2170,7 +2173,10 @@ def union(self, other):
         if len(self) == 0:
             return other._get_consensus_name(self)
 
-        if not is_dtype_equal(self.dtype, other.dtype):
+        # TODO: is_dtype_union_equal is a hack around
+        # 1. buggy joins with duplicates
+        # 2. CategoricalIndex lacking setops
```

Comment:
is this something you want to solve in this PR? Or is there an issue for this?

Reply:
I have a fix for CategoricalIndex set operations,  but it relies on proper support for set ops with duplicates. That ones is more involved. I've added references to the GH issues.